From dd96338dd404d61538e78a10b2e0e119e2e4c2c2 Mon Sep 17 00:00:00 2001
From: Peleg Vigodny <peleg.vigodny@tandemg.com>
Date: Mon, 03 May 2021 16:24:03 +0300
Subject: [PATCH] [SW-41254] gaudi: hbm: check row repair on boot

Perform HBM row repair operation during boot according to row repair
table in flash. If Table is empty, then no row repair is done.

HBM row repair flow is taken from u-boot code:
https://gerrit.habana-labs.com/#/c/91359

Change-Id: I73377f644177465b6e85faa45ccfa23f5f04b725
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/gaudi/hbm.c b/gaudi/hbm.c
index b1ac0ea..7ad3cb7 100644
--- a/gaudi/hbm.c
+++ b/gaudi/hbm.c
@@ -8,6 +8,7 @@
 
 #include <common/pci_ids.h>
 #include <gaudi_blocks.h>
+#include <rma_info.h>
 #include <psoc_global_conf_regs.h>
 
 #include <types.h>
@@ -2483,6 +2484,28 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_UNIT_RST_N);
 }
 
+static void hl_hbm_row_repair_flash_list(void)
+{
+	int err;
+	struct rma_table *table = NULL;
+	size_t size = 0;
+	u8 *p;
+
+	err = rma_table_read(RMA_HBM_ROW_REPLACEMENT_TABLE, &table, &size);
+	if (err || (size == 0))
+		return; /* no row replacement info, nothing to be done */
+
+	for (p = table->data; p < &table->data[size]; p += table->header.entry_size) {
+		struct rma_hbm_row_replacement_entry *entry =
+			(struct rma_hbm_row_replacement_entry *)p;
+
+		gaudi_hbm_row_repair(entry->device, entry->pc, entry->sid,
+				     entry->bank, entry->row);
+	}
+
+	rma_table_free(table, size);
+}
+
 /**
  * gaudi_hbm_init - Initialize HBM controller of the chip and HBM memories
  *		    Scrub the memories.
@@ -2534,6 +2557,8 @@
 	/* Reset repaired failures before AC test */
 	hbm_reset_failures();
 
+	hl_hbm_row_repair_flash_list();
+
 	/* MC BIST (validate Training result and attempt repair) */
 	do {
 		err = mcbist_write_read_full(hbm, 0, true);
diff --git a/gaudi/rma_info.c b/gaudi/rma_info.c
index 119e13a..b991113 100644
--- a/gaudi/rma_info.c
+++ b/gaudi/rma_info.c
@@ -10,7 +10,6 @@
 #include <hl_boot_desc.h>
 #include <string.h>
 #include "uart.h"
-#include "rma_if.h"
 #include "boot_if.h"
 #include "rma_info.h"
 #include "boot_info.h"
@@ -380,3 +379,23 @@
 	if (!ret)
 		boot_info_set(HL_BOOT_RMA_HDR_TABLE, hbm_hdr_table, size);
 }
+
+int rma_table_read(enum rma_tables_type type, struct rma_table **table, size_t *data_len)
+{
+	int ret;
+	u32 size = MAX_INFO_DESC_DATA;
+
+	ret = rma_info_get(type, row_rep_table, &size);
+	if (ret)
+		return ret;
+
+	*table = (struct rma_table *)row_rep_table;
+	*data_len = size - RMA_TABLE_HEADER_SIZE;
+
+	return 0;
+}
+
+void rma_table_free(struct rma_table *table, size_t data_len)
+{
+	memset(table, 0, data_len + RMA_TABLE_HEADER_SIZE);
+}
diff --git a/include/rma_info.h b/include/rma_info.h
index fe0e7ed..c708c4b 100644
--- a/include/rma_info.h
+++ b/include/rma_info.h
@@ -7,7 +7,14 @@
 #ifndef _RMA_INFO_H_
 #define _RMA_INFO_H_
 
+#include <common.h>
 #include <hl_i2c_indirect.h>
+#include <rma_if.h>
+
+struct rma_table {
+	struct table_header header;
+	u8 data[0];
+} __packed;
 
 void preboot_process_last_rma_info(void);
 void rma_info_hbm_par_err_sts(struct hl_i2ci_error_status_hbm_parity *op);
@@ -17,4 +24,17 @@
 void rma_info_hbm_row_replacement(struct hl_i2ci_dram *op);
 void preboot_finalize_rma_tables(void);
 
+/* rma_table_read() - read rma table
+ * @type:	table type
+ * @table:	after successful call, *table will point to an rma table
+ * @data_len:	after successful call, *len will be set to the length of the table data in bytes
+ *
+ * Return:	0:	success
+ *		other:	no info available
+ */
+int rma_table_read(enum rma_tables_type type, struct rma_table **table, size_t *data_len);
+
+/* rma_table_free() - should be called after rma_table_read() */
+void rma_table_free(struct rma_table *table, size_t data_len);
+
 #endif /* _RMA_INFO_H_ */
